cli/upload.py

```diff
@@ -163,45 +163,54 @@ def _start_procs(upload_info: api.UploadInfo, xlsx: str) -> list:
     return procs
 
 
-def _wait_for_upload(procs: list) -> bool:
+def _wait_for_upload(procs: list) -> Optional[str]:
     """
     Waits for all subprocesses and click.echos their stderr streams.
-    Returns bool - if an error has occurred during any if uploads
+    Returns Optional[str] - an error message if an error has occurred during any if uploads
     """
 
     finished = set()
-    errored = False
-    progress_bar = "."
-    while len(finished) != len(procs) and not errored:
-
-        click.secho(f"Uploading files ({len(finished)} / {len(procs)} done)", bold=True)
-        click.echo(progress_bar)
+    error = None
+    prev_errlines = [""] * len(procs)
+    while len(finished) != len(procs) and not error:
         for i, p in enumerate(procs):
-            click.secho(f"(proc {i + 1}) ", nl=False, fg="bright_blue")
-            if i in finished:
-                click.secho("uploaded ", nl=False, fg="green", bold=True)
-                continue
-            else:
-                click.echo("uploading ", nl=False)
+            # start building user feedback for this process
+            message = f"[{len(finished)}/{len(procs)} done] "
+            message += click.style(f"(proc {i + 1}) ", fg="bright_blue")
+
+            # read stderr for this process
+            errline = p.stderr.readline()
 
             if p.poll() != None:
                 finished.add(i)
 
                 if p.returncode != 0:
-                    errored = True
-
-            click.echo(p.args[-2])
+                    message += click.style(
+                        f"!!! upload error !!! ", fg="red", bold=True
+                    )
+                    message += p.args[-2]
+                    click.echo(message)
+                    error = f"{prev_errlines[i]}{errline}"
```

Comment:
not sure what's happening here

Reply:
I'll add a comment that will hopefully clarify.